File: blink-windows/src/Blink.Core.Tests/FlowScoreCalculatorTests.cs
using Blink.Core.Abstractions;
using Blink.Core.FlowDetection;

namespace Blink.Core.Tests;

public class FlowScoreCalculatorTests
{
    [Fact]
    public void NoInput_LowishScore()
    {
        var calc = new FlowScoreCalculator();
        var score = calc.CurrentScore(1000);
        Assert.InRange(score, 0.4, 0.65);
    }

    [Fact]
    public void SustainedTyping_HighScore()
    {
        var calc = new FlowScoreCalculator();
        for (int i = 0; i < 60; i++)
            calc.IngestKeystroke(new KeystrokeEvent(940 + i));
        calc.SetCurrentApp("Code"); // creative app

        var score = calc.CurrentScore(1000);
        Assert.True(score > 0.6, $"Sustained typing in IDE should score high, got {score}");
    }

    [Fact]
    public void FrequentAppSwitching_LowScore()
    {
        var calc = new FlowScoreCalculator();
        for (int i = 0; i < 8; i++)
            calc.RecordAppSwitch(new AppSwitchEvent(750 + i * 30, $"app{i}"));

        var score = calc.CurrentScore(1000);
        Assert.True(score < 0.5, $"Heavy switching should reduce score, got {score}");
    }

    [Fact]
    public void HeavyScrolling_BrowsingScore()
    {
        var calc = new FlowScoreCalculator();
        for (int i = 0; i < 30; i++)
            calc.IngestMouseEvent(new MouseEvent(940 + i * 2, new MouseEventKind.Scroll(10)));

        var score = calc.CurrentScore(1000);
        Assert.True(score < 0.55, $"Scrolling without typing = browsing, got {score}");
    }

    [Fact]
    public void CreativeAppBonus()
    {
        var calc1 = new FlowScoreCalculator();
        var calc2 = new FlowScoreCalculator();
        for (int i = 0; i < 30; i++)
        {
            calc1.IngestKeystroke(new KeystrokeEvent(940 + i * 2));
            calc2.IngestKeystroke(new KeystrokeEvent(940 + i * 2));
        }
        calc1.SetCurrentApp("Code");
        calc2.SetCurrentApp("slack");

        Assert.True(calc1.CurrentScore(1000) > calc2.CurrentScore(1000));
    }

    [Fact]
    public void Reset_ClearsState()
    {
        var calc = new FlowScoreCalculator();
        for (int i = 0; i < 20; i++)
            calc.IngestKeystroke(new KeystrokeEvent(980 + i));
        calc.SetCurrentApp("Code");

        var before = calc.CurrentScore(1000);
        calc.Reset();
        var after = calc.CurrentScore(1000);

        Assert.True(after < before);
    }

    [Fact]
    public void ScoreAlways_0_to_1()
    {
        var calc = new FlowScoreCalculator();
        Assert.InRange(calc.CurrentScore(1000), 0, 1);

        for (int i = 0; i < 50; i++)
        {
            calc.IngestKeystroke(new KeystrokeEvent(950 + i));
            calc.IngestMouseEvent(new MouseEvent(950 + i, new MouseEventKind.Click()));
        }
        Assert.InRange(calc.CurrentScore(1000), 0, 1);
    }
}

File: BlinkCore/Tests/BlinkCoreTests/FlowScoreCalculatorTests.swift
import Foundation
import Testing
@testable import BlinkCore

@Suite("FlowScoreCalculator")
struct FlowScoreCalculatorTests {
    let calculator = FlowScoreCalculator()

    @Test("No input produces low score")
    func noInput() {
        let score = calculator.currentScore(now: 1000)
        // With no input, keystroke=0, mouse=0.5 (neutral), app switch=1.0, window=1.0, context=0.5
        // 0.35*1.0 + 0.25*0 + 0.20*0.5 + 0.10*1.0 + 0.10*0.5 = 0.35+0+0.1+0.1+0.05 = 0.6
        #expect(score < 0.65)
        #expect(score > 0.4)
    }

    @Test("Sustained typing produces high score")
    func sustainedTyping() {
        let now: TimeInterval = 1000

        // Simulate 60 keystrokes over 60 seconds (1 KPS = 60 KPM), steady rhythm
        for i in 0..<60 {
            calculator.ingestKeystroke(KeystrokeEvent(timestamp: now - 60 + Double(i)))
        }

        // Set a creative app
        calculator.setCurrentApp(bundleID: "com.apple.dt.Xcode")

        let score = calculator.currentScore(now: now)
        #expect(score > 0.6, "Sustained typing in Xcode should produce a high flow score, got \(score)")
    }

    @Test("Lots of app switching produces low score")
    func frequentAppSwitching() {
        let now: TimeInterval = 1000

        // 8 app switches in last 5 minutes
        for i in 0..<8 {
            calculator.recordAppSwitch(AppSwitchEvent(
                timestamp: now - 250 + Double(i) * 30,
                appBundleID: "com.app.\(i)"
            ))
        }

        let score = calculator.currentScore(now: now)
        #expect(score < 0.5, "Heavy app switching should reduce flow score, got \(score)")
    }

    @Test("Heavy scrolling with no typing = browsing")
    func scrollingBrowsing() {
        let now: TimeInterval = 1000

        // 30 scroll events, no keystrokes
        for i in 0..<30 {
            calculator.ingestMouseEvent(MouseEvent(
                timestamp: now - 60 + Double(i) * 2,
                kind: .scroll(deltaY: 10)
            ))
        }

        let score = calculator.currentScore(now: now)
        #expect(score < 0.55, "Scrolling without typing should indicate browsing, got \(score)")
    }

    @Test("Creative app bonus")
    func creativeAppBonus() {
        let now: TimeInterval = 1000

        // Same typing pattern
        for i in 0..<30 {
            calculator.ingestKeystroke(KeystrokeEvent(timestamp: now - 60 + Double(i) * 2))
        }

        calculator.setCurrentApp(bundleID: "com.apple.dt.Xcode")
        let xcodeScore = calculator.currentScore(now: now)

        let calculator2 = FlowScoreCalculator()
        for i in 0..<30 {
            calculator2.ingestKeystroke(KeystrokeEvent(timestamp: now - 60 + Double(i) * 2))
        }
        calculator2.setCurrentApp(bundleID: "com.tinyspeck.slackmacgap")
        let slackScore = calculator2.currentScore(now: now)

        #expect(xcodeScore > slackScore, "Xcode (\(xcodeScore)) should score higher than Slack (\(slackScore))")
    }
}
